tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/Scope.java

```diff
@@ -211,46 +248,81 @@ private Scope(
    * @return a new scope with the provided control dependencies
    */
   public Scope withControlDependencies(Iterable<Op> controls) {
-    for (Op control : controls) {
+    return withControlDependencyOps(
+        StreamSupport.stream(controls.spliterator(), false)
+            .map(Op::op)
+            .collect(Collectors.toList()));
+  }
+
+  /**
+   * Returns a new scope where added operations will have the provided control dependencies.
+   *
+   * <p>Ops created with this scope will have a control edge from each of the provided controls. All
+   * other properties are inherited from the current scope.
+   *
+   * <p>Init ops will never be used as control dependencies, they are assumed to be created during
+   * session initialization.
+   *
+   * @param controls control dependencies for ops created with the returned scope
+   * @return a new scope with the provided control dependencies
+   */
+  public Scope withControlDependencyOps(Iterable<Operation> controls) {
+    for (Operation control : controls) {
       env.checkInput(control);
     }
-    return new Scope(env, nameScope, controls, deviceSpec);
+
+    return new Scope(env, nameScope, controls, deviceSpec, isInit);
   }
 
   /**
    * Applies device specification and adds each Operand in controlDependencies as a control input to
    * the provided builder.
    *
+   * <p><b>Should only be used from {@link OperationBuilder} implementations</b>
+   *
    * @param builder OperationBuilder to add control inputs and device specification to
    */
   public OperationBuilder apply(OperationBuilder builder) {
     builder.setDevice(deviceSpec.toString());
-    return applyControlDependencies(builder);
+    for (Operation control : controlDependencies) {
+      if (isInit || !env.isInitOp(control)) {
+        builder.addControlInput(control);
+      }
+    }
+    return builder;
   }
 
   /**
-   * Adds each Operand in controlDependencies as a control input to the provided builder.
+   * Handle op creation, like registering it as an init op if the scope is init.
    *
-   * @param builder OperationBuilder to add control inputs to
+   * <p><b>FOR INTERNAL USE ONLY</b>
    */
-  public OperationBuilder applyControlDependencies(OperationBuilder builder) {
-    for (Op control : controlDependencies) {
-      builder = builder.addControlInput(control.op());
+  public void onOpCreated(Operation op) {
+    if (isInit) {
+      env.registerInitOp(op);
     }
-    return builder;
   }
 
   private final ExecutionEnvironment env;
-  private final Iterable<Op> controlDependencies;
+  private final Iterable<Operation> controlDependencies;
   private final NameScope nameScope;
   private final DeviceSpec deviceSpec;
+  private final boolean isInit;
 
   /** Returns device string from the scope. */
   public String getDeviceString() {
     return deviceSpec.toString();
   }
+
+  /** Get whether this scope is building init ops. */
+  public boolean isInit() {
+    return isInit;
+  }
```

Comment:
Nit: order of `private` and `public` endpoints are getting a bit messy, can you move this method (and `getDeviceString()` while at it) before the private fields?